nexus/src/app/instance.rs

```diff
@@ -267,13 +274,29 @@ impl super::Nexus {
         instance_lookup: &lookup::Instance<'_>,
         params: params::InstanceMigrate,
     ) -> UpdateResult<db::model::Instance> {
-        let (.., authz_instance) =
-            instance_lookup.lookup_for(authz::Action::Modify).await?;
+        let (.., authz_instance, db_instance) =
+            instance_lookup.fetch_for(authz::Action::Modify).await?;
+
+        if db_instance.runtime().state.0 != InstanceState::Running {
+            return Err(Error::invalid_request(
+                "instance must be running before it can migrate",
+            ));
+        }
+
+        if db_instance.runtime().sled_id == params.dst_sled_id {
+            return Err(Error::invalid_request(
+                "instance is already running on destination sled",
+            ));
+        }
+
+        if db_instance.runtime().migration_id.is_some() {
+            return Err(Error::unavail("instance is already migrating"));
+        }
 
         // Kick off the migration saga
         let saga_params = sagas::instance_migrate::Params {
             serialized_authn: authn::saga::Serialized::for_opctx(opctx),
-            instance_id: authz_instance.id(),
+            instance: db_instance,
```

Comment:
This is a good idea. There's another place in the create saga where this is relevant (create an instance without starting it; saga writes the Stopped state back to CRDB; now the instance is user-visible and not Creating; user starts it; saga step gets replayed; now Nexus is writing state that sled agent owns) and where using a serialized instance record would help (by stopping a replayed saga step from picking up a more-advanced generation number than the one the instance was created with). I've filed #2945 to track.

Reply:
That said, since this PR is already chest-deep in migration saga changes, it probably makes sense to go ahead and have the saga output the expected record as part of this PR (and just have 2945 deal with the instance creation piece). I'll work on that.

Reply:
Sooo... I'm actually not sure what to return here. Naively, I'd expect that if I call the `migrate` endpoint and get an `Instance` back, that `Instance` will say it's `Migrating`. But the saga never actually puts the `Migrating` state into anything--that's written back by the source Propolis once the migration actually begins--so there's no obvious node output that we could use as a return value here. The saga could synthesize one, but it would be a total fiction, and I don't think it tells the caller anything they don't know already (if the request to migrate succeeded, then you already know the system is at least going to attempt a migration).

This case seems to me to differ from the others where Nexus returns an `Instance`, in the following way: if you ask to stop or reboot an instance, Propolis and sled agent guarantee that (absent an unexpected failure) your instance *will eventually* do the thing you asked it to do--it's on the relevant state change queue. When you ask to migrate, though, the migration could fail, and indeed it might never even begin (e.g. if the source instance shuts down just before the destination starts to talk to it).

It almost seems to me like the migration endpoint shouldn't return an `Instance` at all and should instead return information about the *migration attempt itself*. This provides the caller with some new information--the migration ID--that it could use in subsequent calls (in a hypothetical world where Nexus could track and report on migration statuses).

@luqmana @jordanhendricks @smklein Any thoughts on all of this? I'm unsure how much to value consistency between `instance_migrate` and other instance APIs but don't want to punt on this if y'all have strong feelings about how the instance API should be designed.

Reply:
I think your analysis there makes sense and we could not return the `Instance` from the migrate endpoint. Returning an ID if it was able to successfully kick one off or an error otherwise seems reasonable.

Reply:
I've filed #2948 to track all this. (We could return the migration ID right away, but it's not useful because there are currently no Nexus endpoints to query migration statuses as such; some of that work will come in the next PR, but there's a bunch of Nexus stuff we'd have to do as well, which the new issue will track.)